### PR TITLE
[EN TasksMode] Suppressing specific behaviours under TasksMode

### PR DESCRIPTION
**Pull request by Neelisha-saxena**
Under Tasksmode following scenarios are implemented.
 
2. suppress number with s e.g., 90s from getting classified as date range.
3. Remove Holidays from holiday list, celebration date of which are region specific example Independence Day.
4. Suppress time expression like 1ampm etc.
5. Suppress time-period like Q1, quarter etc.

### CODE REVIEW

File: .NET/Microsoft.Recognizers.Definitions.Common/English/DateTimeDefinitions.cs [hunk 1]
```diff
@@ -239,6 +239,7 @@ public static class DateTimeDefinitions
...
       public const string HolidayList4 = @"(?<holiday>ramad(h)?an|ram(a)?zan|ramathan|eid al(-|\s+)adha|eid al(-|\s+)azha|eidul(-|\s+)azha|feast of the sacrifice|(islamic|arabic|hijri) new year|eid al(-|\s+)fitr|festival of breaking the fast)";
       public static readonly string HolidayRegex = $@"\b(({StrictRelativeRegex}\s+({HolidayList1}|{HolidayList2}|{HolidayList3}|{HolidayList4}))|(?<holidayWeekend>((the\s+)?weekend\s+of\s+)({HolidayList1}|{HolidayList2}|{HolidayList3}|{HolidayList4})(\s+((of\s+)?({YearRegex}|{RelativeRegex}\s+year)))?)|(({HolidayList1}|{HolidayList2}|{HolidayList3}|{HolidayList4})((?<holidayWeekend>(\s+weekend)(\s+((of\s+)?({YearRegex}|{RelativeRegex}\s+year)))?)|(\s+(of\s+)?({YearRegex}|{RelativeRegex}\s+year)(?<holidayWeekend>\s+weekend)?))?))\b";
+      public static readonly string HolidayListSupress = $@"(?<holiday>(?:independence|teacher'?s|youth|children|girls)\s+day)|{HolidayList4}";
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Definitions.Common/English/DateTimeDefinitions.cs:242]: This should be named according to your mode. And why include most Islamic holidays? In this particular case, it's better to repeat the holidays here, in case future changes in list 4 add other holidays.

File: .NET/Microsoft.Recognizers.Text.DateTime/Dutch/Extractors/DutchMergedExtractorConfiguration.cs [hunk 2]
```diff
@@ -64,6 +64,8 @@ public class DutchMergedExtractorConfiguration : BaseDateTimeOptionsConfiguratio
             new Regex(DateTimeDefinitions.SingleAmbiguousTermsRegex, RegexFlags),
         };
 
+        public static readonly Regex[] TasksSkipMergeRegexes = System.Array.Empty<Regex>();
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Dutch/Extractors/DutchMergedExtractorConfiguration.cs:67]: Maybe rename to TasksModeMentionFilters?

File: .NET/Microsoft.Recognizers.Text.DateTime/English/Extractors/EnglishMergedExtractorConfiguration.cs [hunk 3]
```diff
@@ -64,6 +64,38 @@ public class EnglishMergedExtractorConfiguration : BaseDateTimeOptionsConfigurat
 
             // (the)? (day|week|month|year)
             new Regex(DateTimeDefinitions.SingleAmbiguousTermsRegex, RegexFlags),
+
+        };
+
+        // Skip Regexes under TasksMode
+        public static readonly Regex[] TasksSkipMergeRegexes =
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/English/Extractors/EnglishMergedExtractorConfiguration.cs:71]: Please surround creation/load with a check for the proper mode being set. If mode is not set, this should not be created.
Also, is the set of regexes in this array always the same for all languages? If so, we should declare it only in one place instead of all configs.

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseMergedDateTimeExtractor.cs [hunk 4]
```diff
@@ -276,6 +276,23 @@ private bool ShouldSkipFromToMerge(ExtractResult er)
             return config.FromToRegex.IsMatch(er.Text);
         }
 
+        private bool SkipRegexTasksMode(ExtractResult er)
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseMergedDateTimeExtractor.cs:279]: Nit: Rename accoridng to previous suggestion.

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseMergedDateTimeExtractor.cs [hunk 5]
```diff
@@ -276,6 +276,23 @@ private bool ShouldSkipFromToMerge(ExtractResult er)
...
+
+                if (match.Success)
+                {
+                    return true;
+
+                }
+            }
+
+            return false;
+
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseMergedDateTimeExtractor.cs:293]: Nit: Remove unnecessary blank line.

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseMergedDateTimeExtractor.cs [hunk 6]
```diff
@@ -276,6 +276,23 @@ private bool ShouldSkipFromToMerge(ExtractResult er)
...
+        private bool SkipRegexTasksMode(ExtractResult er)
+        {
+            foreach (var regex in this.config.TasksSkipMergeRegexes)
+            {
+                var match = regex.Match(er.Text);
+
+                if (match.Success)
+                {
+                    return true;
+
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseMergedDateTimeExtractor.cs:288]: Nit: Remove unnecessary blank line.

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/IMergedExtractorConfiguration.cs [hunk 7]
```diff
@@ -36,6 +36,8 @@ public interface IMergedExtractorConfiguration : IDateTimeOptionsConfiguration
 
         IEnumerable<Regex> TermFilterRegexes { get; }
 
+        IEnumerable<Regex> TasksSkipMergeRegexes { get; }
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Extractors/IMergedExtractorConfiguration.cs:39]: Add a code commen here explaining the purpose of this and making it clear it's only used if the specific mode is enabled.

File: Specs/DateTime/English/DateTimeModel.json [hunk 8]
```diff
@@ -23962,5 +23962,304 @@
         }
       }
     ]
+  },
+  {
+    "Input": "schedule me a 1:1 meeting",
```
Comment (tellarin) [inline Specs/DateTime/English/DateTimeModel.json:23967]: This is actually incorrect and should be discarded by an amboguity filter. Weird.
Test specs document the desired behaviour. This specific spec is incorrect and should be removed. Plus the case should be properly handled (it can be in a separate PR).

File: Specs/DateTime/English/DateTimeModel.json [hunk 9]
```diff
@@ -23962,5 +23962,304 @@
...
+              "timex": "T13:01",
+              "type": "time",
+              "value": "13:01:00"
+            }
+          ]
+        }
+      }
+    ]
+  },
+  {
```
Comment (tellarin) [inline Specs/DateTime/English/DateTimeModel.json:23996]: This one should be modified to remove the 1:1 extraction once functionality is corrected.

File: Specs/DateTime/English/DateTimeModel.json [hunk 10]
```diff
@@ -23962,5 +23962,304 @@
...
+              "type": "date",
+              "value": "2016-11-14"
+            }
+          ]
+        }
+      }
+    ]
+  },
+  {
+    "Input": "schedule me a 1:1 on saturday",
```
Comment (tellarin) [inline Specs/DateTime/English/DateTimeModel.json:24046]: Same and so on.

File: Specs/DateTime/English/DateTimeModel.json [hunk 11]
```diff
@@ -23962,5 +23962,304 @@
...
+              "value": "2023-11-14"
+            }
+          ]
+        }
+      }
+    ]
+  },
+  {
+    "Input": "Listen to 1990s music",
+    "Comment": "Taskmode will supress number with s (90s) from getting be categorised as daterange",
```
Comment (tellarin) [inline Specs/DateTime/English/DateTimeModel.json:24130]: No need for such comment in the spec for the default mode if the behaviour in this mode is correct/desired.

File: Specs/DateTime/English/DateTimeModelTasksMode.json [hunk 12]
```diff
@@ -88,5 +88,322 @@
...
+    "Input": "Listen to 1990s music",
+    "Comment": "Taskmode will supress number with s (90s) from getting be categorised as daterange",
+    "Context": {
+      "ReferenceDateTime": "2018-05-29T00:00:00"
+    },
+    "NotSupported": "javascript, python, java",
+    "Results": []
+  },
+  {
+    "Input": "Who are us presidents in 1990 s.",
```
Comment (tellarin) [inline Specs/DateTime/English/DateTimeModelTasksMode.json:193]: Should there be the whitespace before the 's'? 
Also, ideally there should be a 1-1 map of specs in the new mode and their versions in the default mode. Is this the only missing? Why?

**Review by tellarin — CHANGES_REQUESTED** [review]

File: Specs/DateTime/English/DateTimeModel.json (hunk 10 above)
Comment (aitelint) [reply Specs/DateTime/English/DateTimeModel.json:24046]: it is extracted as time, maybe we could modify MinuteRegex in Base-DateTime from '[0-5]?\d' to '[0-5]\d' ? So the problem will be solved in all languages (the tests results are not affected).

Comment (tellarin) [reply Specs/DateTime/English/DateTimeModel.json:24356]: @aitelint that seems reasonable if the minute regex is not used in some other case that'd break. Can you PR the change?

File: .NET/Microsoft.Recognizers.Definitions.Common/English/DateTimeDefinitions.cs (hunk 1 above)
Comment (Neelisha-saxena) [reply .NET/Microsoft.Recognizers.Definitions.Common/English/DateTimeDefinitions.cs:246]: Included Islamic holidays as its celebration date vary according to geographic location.

File: .NET/Microsoft.Recognizers.Text.DateTime/English/Extractors/EnglishMergedExtractorConfiguration.cs (hunk 3 above)
Comment (Neelisha-saxena) [reply .NET/Microsoft.Recognizers.Text.DateTime/English/Extractors/EnglishMergedExtractorConfiguration.cs:71]: For creation/load "TasksSkipMergeRegexes", I have followed similar convention as it's done under CalendarMode for creation for "TermFilterRegexes".

Comment (Neelisha-saxena) [reply .NET/Microsoft.Recognizers.Text.DateTime/English/Extractors/EnglishMergedExtractorConfiguration.cs:71]: TasksSkipMergeRegexes :  This can be different under different languages, right now we are not sure how this filter will work for other languages, so it's assigned as empty array

Comment (Neelisha-saxena) [reply .NET/Microsoft.Recognizers.Text.DateTime/English/Extractors/EnglishMergedExtractorConfiguration.cs:71]: > For creation/load "TasksSkipMergeRegexes", I have followed similar convention as it's done under CalendarMode for creation for "TermFilterRegexes".

Comment (Neelisha-saxena) [reply .NET/Microsoft.Recognizers.Text.DateTime/English/Extractors/EnglishMergedExtractorConfiguration.cs:71]: > Please surround creation/load with a check for the proper mode being set. If mode is not set, this should not be created. Also, is the set of regexes in this array always the same for all languages? If so, we should declare it only in one place instead of all configs.

For creation/load "TasksSkipMergeRegexes", I have followed similar convention as it's done under CalendarMode for creation for "TermFilterRegexes".

Comment (Neelisha-saxena) [reply .NET/Microsoft.Recognizers.Text.DateTime/English/Extractors/EnglishMergedExtractorConfiguration.cs:71]: > Please surround creation/load with a check for the proper mode being set. If mode is not set, this should not be created. Also, is the set of regexes in this array always the same for all languages? If so, we should declare it only in one place instead of all configs.

TasksSkipMergeRegexes :  This can be different under different languages, right now we are not sure how this filter will work for other languages, so it's assigned as empty array

File: .NET/Microsoft.Recognizers.Text.DateTime/Dutch/Extractors/DutchMergedExtractorConfiguration.cs (hunk 2 above)
Comment (Neelisha-saxena) [reply .NET/Microsoft.Recognizers.Text.DateTime/Dutch/Extractors/DutchMergedExtractorConfiguration.cs:67]: Renamed TasksSkipMergeRegexes to TasksModeMentionFilters everywhere in code.

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseMergedDateTimeExtractor.cs (hunk 4 above)
Comment (Neelisha-saxena) [reply .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseMergedDateTimeExtractor.cs]: Renamed SkipRegexTasksMode to TasksModeFilters.

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseMergedDateTimeExtractor.cs (hunk 6 above)
Comment (Neelisha-saxena) [reply .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseMergedDateTimeExtractor.cs]: removed blank line.

File: Specs/DateTime/English/DateTimeModel.json (hunk 9 above)
Comment (Neelisha-saxena) [reply Specs/DateTime/English/DateTimeModel.json]: Changes integrated

File: Specs/DateTime/English/DateTimeModel.json (hunk 11 above)
Comment (Neelisha-saxena) [reply Specs/DateTime/English/DateTimeModel.json]: removed unnecessary comment tag.

File: Specs/DateTime/English/DateTimeModelTasksMode.json (hunk 12 above)
Comment (Neelisha-saxena) [reply Specs/DateTime/English/DateTimeModelTasksMode.json:193]: >Since the following example, "Who are us presidents in the 1990s," already existed in default mode, I simply copied it and modified the outcome under tasksmode.

File: .NET/Microsoft.Recognizers.Text.DateTime/English/Extractors/EnglishMergedExtractorConfiguration.cs [hunk 13]
```diff
@@ -64,6 +64,38 @@ public class EnglishMergedExtractorConfiguration : BaseDateTimeOptionsConfigurat
 
             // (the)? (day|week|month|year)
             new Regex(DateTimeDefinitions.SingleAmbiguousTermsRegex, RegexFlags),
+
+        };
+
+        // Skip Regexes under TasksMode
+        public static readonly Regex[] TasksModeMentionFilters =
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/English/Extractors/EnglishMergedExtractorConfiguration.cs:71]: As mentioned, I'd prefer is this is not automatically created always. Only if the mode is enabled.
Perhaps a flag check in the constructor and then call a build method?

File: Specs/DateTime/English/DateTimeModelTasksMode.json [hunk 14]
```diff
@@ -88,5 +88,346 @@
...
+              "type": "date",
+              "value": "2016-11-14"
+            }
+          ]
+        }
+      }
+    ]
+  },
+  {
+    "Input": "schedule me a 1:1 on saturday",
```
Comment (tellarin) [inline Specs/DateTime/English/DateTimeModelTasksMode.json:131]: Can you add these with 1:1 in the default mode to show the 1:1 is not extracted, but the other mentions like Saturday are?

**Review by tellarin — CHANGES_REQUESTED** [review]

File: Specs/DateTime/English/DateTimeModelTasksMode.json (hunk 14 above)
Comment (Neelisha-saxena) [reply Specs/DateTime/English/DateTimeModelTasksMode.json:131]: I have updated DatetimeModel.json with more examples, where 1:1 is not extracted but other mentions of datetime gets extracted.

File: .NET/Microsoft.Recognizers.Text.DateTime/English/Extractors/EnglishMergedExtractorConfiguration.cs (hunk 13 above)
Comment (Neelisha-saxena) [reply .NET/Microsoft.Recognizers.Text.DateTime/English/Extractors/EnglishMergedExtractorConfiguration.cs:72]: I have made changes, now when tasks Mode sets to true, then only TasksModeMentionFilters will get created.

Comment (tellarin) [reply .NET/Microsoft.Recognizers.Text.DateTime/English/Extractors/EnglishMergedExtractorConfiguration.cs:72]: Can you delete this commented our static block?

File: .NET/Microsoft.Recognizers.Text.DateTime/English/Extractors/EnglishMergedExtractorConfiguration.cs [hunk 15]
```diff
@@ -96,6 +129,39 @@ public EnglishMergedExtractorConfiguration(IDateTimeOptionsConfiguration config)
                 SinceRegex = SinceRegexExp;
             }
 
+            if ((config.Options & DateTimeOptions.TasksMode) != 0)
+            {
+                TasksModeMentionFilters = new List<Regex>
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/English/Extractors/EnglishMergedExtractorConfiguration.cs:134]: Please move this init into it's own protected method and fix code indentation. It's easier to read the constructor code.

File: .NET/Microsoft.Recognizers.Text.DateTime/English/Extractors/EnglishMergedExtractorConfiguration.cs [hunk 16]
```diff
@@ -177,5 +243,9 @@ public EnglishMergedExtractorConfiguration(IDateTimeOptionsConfiguration config)
         StringMatcher IMergedExtractorConfiguration.SuperfluousWordMatcher => SuperfluousWordMatcher;
 
         bool IMergedExtractorConfiguration.CheckBothBeforeAfter => DateTimeDefinitions.CheckBothBeforeAfter;
+
+        // IEnumerable<Regex> IMergedExtractorConfiguration.TasksModeMentionFilters => TasksModeMentionFilters;
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/English/Extractors/EnglishMergedExtractorConfiguration.cs:247]: Remove commented out code.

**Review by tellarin — CHANGES_REQUESTED** [review]
Only very minor issues left. We can merge it soon.
Thanks.

File: .NET/Microsoft.Recognizers.Text.DateTime/English/Extractors/EnglishMergedExtractorConfiguration.cs (hunk 13 above)
Comment (Neelisha-saxena) [reply .NET/Microsoft.Recognizers.Text.DateTime/English/Extractors/EnglishMergedExtractorConfiguration.cs]: deleted comment block.

**Review by tellarin — APPROVED** [review]

**Comment by tellarin** [comment]
@Neelisha-saxena Please resolve the conflicts to merge. Thanks.

**Comment by Neelisha-saxena** [comment]
> @Neelisha-saxena Please resolve the conflicts to merge. Thanks.

resolved merge conflicts
